compiler/rustc_infer/src/infer/error_reporting/nice_region_error/different_lifetimes.rs

```diff
@@ -166,59 +166,55 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
-                    {
-                        let (suggestion_param_name, introduce_new) = generics
-                            .params
-                            .iter()
-                            .find(|p| matches!(p.kind, GenericParamKind::Lifetime { .. }))
-                            .and_then(|p| self.tcx().sess.source_map().span_to_snippet(p.span).ok())
-                            .map(|name| (name, false))
-                            .unwrap_or_else(|| ("'a".to_string(), true));
-
-                        let mut suggestions = vec![
-                            if let hir::LifetimeName::Underscore = lifetime_sub.name {
-                                (lifetime_sub.span, suggestion_param_name.clone())
-                            } else {
-                                (
-                                    lifetime_sub.span.shrink_to_hi(),
-                                    suggestion_param_name.clone() + " ",
-                                )
-                            },
-                            if let hir::LifetimeName::Underscore = lifetime_sup.name {
-                                (lifetime_sup.span, suggestion_param_name.clone())
-                            } else {
-                                (
-                                    lifetime_sup.span.shrink_to_hi(),
-                                    suggestion_param_name.clone() + " ",
-                                )
-                            },
-                        ];
-
-                        if introduce_new {
-                            let new_param_suggestion = match &generics.params {
-                                [] => (generics.span, format!("<{}>", suggestion_param_name)),
-                                [first, ..] => (
-                                    first.span.shrink_to_lo(),
-                                    format!("{}, ", suggestion_param_name),
-                                ),
-                            };
-
-                            suggestions.push(new_param_suggestion);
-                        }
-
-                        err.multipart_suggestion(
-                            "consider introducing a named lifetime parameter",
-                            suggestions,
-                            Applicability::MaybeIncorrect,
-                        );
-                        err.note(
-                            "each elided lifetime in input position becomes a distinct lifetime",
-                        );
+
+                    let generics = match self.tcx().hir().get(hir_id) {
+                        hir::Node::Item(&hir::Item {
+                            kind: hir::ItemKind::Fn(_, ref generics, ..),
+                            ..
+                        })
+                        | hir::Node::TraitItem(&hir::TraitItem { ref generics, .. })
+                        | hir::Node::ImplItem(&hir::ImplItem { ref generics, .. }) => generics,
```

Comment:
We should handle the `impl` case specifically with extra wording in the suggestion to let people know that they'll have to modify the `trait` as well. This is likely not a problem in practice, but I'd like us to be thorough and avoid confusing people.

It might be enough to just have a boolean flag that you use to modify the suggestion description with extra wording.

Reply:
Good call. I'll make that change.

Reply:
How about this for `impl`? 

```
help: consider introducing a named lifetime parameter and update trait if needed
   |
32 |     fn foo<'a>(&'a self, x: &i32, z: &'a i32) -> Option<&i32> {
   |           ++++  ++                    ++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0623`.
```

Reply:
Ideally we would have something like "consider introducing a named lifetime parameter and update trait `TraitName` as necessary", by getting the `TraitName` from the parent DefId, but that can be left as follow up work.